### ISSUE TITLE
Avoid subtracting `Duration` from `Instant`

### ISSUE THREAD

**Issue by CryZe** [rust-lang/rust-clippy#13020]
### What it does

It's a major footgun to subtract `Duration`s from `Instant`s . On a bunch of targets `Instant` is quite close to the "beginning of time" / "0" when the application starts. `Instant` is also an unsigned type, so subtracting anything from it, can very easily underflow it and panic. You should instead prefer subtracting from `Durations` that result from `Instant`s, or possibly even avoid that, cause that can also panic easily. Ideally you rewrite your code to use additions instead.

### Advantage

- Code is more portable
- Less likely to panic

### Drawbacks

- Doing the operation on the `Duration`s can result in one more operation in rare cases

### Example

```rust
Instant::now() - waiting_time >= previous_instant
```

Could be written as:

```rust
Instant::now() - previous_instant >= waiting_time 
```

or

```rust
previous_instant.elapsed() >= waiting_time
```

Though tbh, I'm not sure it could / should even show any such suggestion, as the solution space is too large.

**Comment by Jacherr** [comment]
This seemingly already exists as a pendantic lint: https://rust-lang.github.io/rust-clippy/master/#/unchecked_duration_subtraction

@rustbot release-assignment

**Comment by CryZe** [comment]
TIL, yeah so I guess this can be closed then.

**Comment by CryZe** [comment]
Is pedantic maybe too nieche of a lint group though? You basically should never subtract from Instant is what I think. In fact I was contemplating opening an issue directly in the Rust repo to possibly even deprecate it.

**Comment by Jacherr** [comment]
Might as well make this change and see what people think. On the whole I agree.

@rustbot claim

**Comment by Jacherr** [comment]
FWIW, this lint *does* differ in what you suggested by suggesting the use of `checked_div` instead of rewriting to use addition.

**Comment by CryZe** [comment]
Yeah there's definitely different solutions, but in general it should probably be avoided. Though there's likely not a good way to recover from the checked operation failing. Not subtracting at all in case it fails is probably the wrong behavior. Rewriting in terms of an addition is generally the safer solution, just not one that clippy can offer automatically probably.

**Comment by Alexendoo** [comment]
See also https://github.com/rust-lang/rust-clippy/issues/10061 and the linked zulip discussions. Becoming on by default again is not impossible but the help message would have to be improved to suggest avoiding the operation rather than replacing it with a slightly worse one

I think a concrete example for the description could also help, when I read things like "certain platforms" I assume it means uncommon ones but that may not be the case
